File: Data Structures & Algorithms/max-area-of-island/submission-1.java
class Solution {
    private int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        int rows = grid.length;
        int cols = grid[0].length;

        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if (grid[i][j] == 1){
                    maxArea = Math.max(maxArea, dfs(grid, i, j));
                }
            }
        }
        return maxArea;
    }
    private int dfs(int[][] grid, int r, int c){
        int rows = grid.length;
        int cols = grid[0].length;
        if (r < 0 || c < 0 || r >= rows || c >= cols || grid[r][c] == 0){
            return 0;
        }
        int res = 1;
        grid[r][c] = 0;
        for(int[] dir : directions){
            res += dfs(grid, r + dir[0] , c + dir[1]);
        }
        return res;
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = {} # key => (i, val) , val = amount
        if len(nums) == 1:
            return nums[0]
        def dfs(i, flag):
            if i >= len(nums) or (flag and i == len(nums) - 1):
                return 0

            if (i, flag) in dp:
                return dp[(i, flag)]

            dp[(i, flag)] = max(dfs(i + 1, flag), nums[i] + dfs(i + 2, flag))
            return dp[(i, flag)]
        return max(dfs(0, True), dfs(1, False))